### PR TITLE
Refactor heap and cache outers

### PR DESCRIPTION
**Pull request by liufengyun**
Some refactoring:

- Refactor heap 
- Cache outers
- Use opaque types for domain definitions

### CODE REVIEW

**Comment by liufengyun** [comment]
@olhotak @EnzeXing This PR will make the experiments easier.

Allowing non-hot as arguments to primary constructors would need to remember the value of the parameters in the heap. Thus reintroduction of the heap and addresses is inevitable. It also helps with performance as it enables more caching.

**Comment by olhotak** [comment]
In extending the abstract heap to reason about multiple objects, a key concern is the design of addresses. Since in this case, we want must information (sets of fields that must have been assigned), we need each abstract address to represent (at most) one single concrete object. `ThisRef(C)` is a good address in that it represents the unique object currently pointed to by `this`. All-objects-of-class-B is a bad address in that there may be multiple concrete objects of class B at a time.

Since we want the heap to reason about objects whose constructors we are analyzing, perhaps the design of addresses could be based on the stack of constructors that we analyze. For example, for the program
```
class A {
  val b = new B(A.this)
}
```
the call stack in our analysis might look like this:
```
1: constructor of B called from
0: constructor of A
```
Then our addresses could be something like `ThisRef(0)` and `ThisRef(1)`, referring to the values of `this` in the two frames of the call stack. (Only constructor frames would give rise to addresses. Frames for methods would not because a method call does not create a new object.)

**Comment by liufengyun** [comment]
@olhotak Thanks for the feedback. This PR is essentially the same as the master --- the heap enables more caching. It would be good to motivate the proposal above with concrete examples which are related to concerns about _expressiveness_, _soundness_ or _termination_. Let's discuss more tomorrow.

**Comment by olhotak** [comment]
I thought some more about how to design the `Heap` and `Addr`.

We have two analyses, `init` and `eval`. Only `init` mutates the fields of the current `Objekt`. `eval` reads them when it encounters `ThisRef`, but it doesn't change them. Also, `init` mutates the fields of only one single `Objekt`.

This suggests keeping a single `Objekt` with mutable fields.

But then we want to leak non-hot values to methods (method parameters), constructors, and to local variables of methods. For this, we need to keep track of which method parameters, constructor parameters (fields), and local variables have non-hot values. So we need an Environment. An Environment should be a map from parameters/variables to `Value`s. Also, outer, outer.outer, outer.outer.outer.... are like variables, so also can be keys of the Environment. Perhaps this Environment is what the generalized `Heap` was intended to be. And `Addr` were intended to be its keys. But parameters/variables/outer are not abstractions of concrete addresses, so this is confusing.

Unlike `Objekt`/`Heap`, such an Environment is immutable in that it is passed as input to `eval` and to `init`, but does not have mutable fields. The `Objeckt`/`Heap` is passed to `init` for reading and writing. `eval` needs to read it but not mutate it.

Then perhaps such an Environment can replace `thisV`. Although the type of `thisV` is a `Value`, it can really only be `ThisRef(C)` or `Warm(C, outer: Value)`, indicating that outer is hot or outer has some other value, respectively. If we generalize to allow variables/parameters other than outer to be non-hot, then the temperature of outer can be recorded together with the temperature of other variables/parameters, rather than on `thisV`.

I find `Warm` a bit confusing because it is not used in general for warm values (all fields assigned with not necessarily hot values) but only for the special case of a `ThisRef` whose outer is non-hot.

After this, one possible further precision improvement could be to extend the `Value` lattice with something between `Warm` and `Cold`, an object with some fields assigned and others not, much like `Objekt`, but as a `Value`, and thus with immutable fields. It could be an immutable view of `Objekt`. This would be useful if we want to allow leaking a `this` with not all fields initialized into methods/constructors and we want to allow accessing specific fields of that `this` in those methods/constructors.

**Comment by liufengyun** [comment]
Thank you @olhotak for the many good points, I agree with many of them.

At the high-level, many styles of analysis are equivalent in essence,
however they make a big difference in engineering qualities. E.g., the
master (based on abstract definitional interpreters, ADI) and the
type-and-effect system are both receiver-polyvariant and have same the
time complexity, but the former has better engineering qualities:

- the length restriction is gone
- simplicity and maintainability
- extensibility

In addition to local reasoning, another aspect that is heavily
employed in both the previous type-and-effect analysis and the current
ADI is _heap monotonicity_. Thanks to it, in the ADI version, we can
cache without making the heap part of the key and get away from
fixed-point computation. That is also the reason why we don't need to
thread through the heap, but instead use it as a _global_ cache.  It's
just one line of difference in code, but it's of far-reaching
importance.

Note that it's not enough that `eval` does not mutate the heap.  Heap
monotonicity in addition ensures that if we have a more initialized
heap, the result of `eval` for the same expression and environment
should still produce a sound abstraction for the concrete value of the
expression computed from the old heap. That enables caching without
heap as part of the key and without fix-point computation.

I conjecture many analysis lie in the category of heap-monotonistic
analysis, thus can be implemented in the same style as initialization
analysis. For example, in type analysis, mutation does not play a role
in type inference --- the type of the initializer of a field determines
its type.

> But then we want to leak non-hot values to methods (method parameters), constructors, and to local variables of methods. For this, we need to keep track of which method parameters, constructor parameters (fields), and local variables have non-hot values. So we need an Environment. An Environment should be a map from parameters/variables to Values.

Yes, that's correct. Introduction of environment is a natural step. We
need to be careful with the introduction to ensure soundness,
termination and good performance. For example, `Fun` may now capture
the environment, finitizing measures must be taken. The same holds for
local classes. BTW, local variables and parameters can be handled
differently, we can discuss the detail when we reach the point.

As a first step, to support non-hot values to primary constructors,
actually we can avoid introducing environments, which will be a
straight-forward extension based on this PR.

To support non-hot values for secondary constructors, we may still
keep the abstract domain for functions and (local) classes the
same. But care must be taken for closures and local classes in
secondary constructors, which should be rare in practice but still
possible.

> Also, outer, outer.outer, outer.outer.outer.... are like variables, so also can be keys of the Environment.

Conceptually, it makes sense. However, it ignores an important
semantic property of inner classes if the outer is always a stable
path: i.e. the outers of super-classes are all determined by the
immediate outer of the concrete class of an object. That property
enables a more concise representation of outers. That is also the
reason why the abstract address `Warm` only contains one outer as part
of the key to the abstract heap (summary).

> Then perhaps such an Environment can replace `thisV`. 

I agree, conceptually it's possible to treat `this` as the parameter 0
of methods. However, there is no engineering benefits in doing
so. Even in coarse-grained analysis like type systems, it handles
`this` differently from parameters. In our setting, the analysis is
receiver-sensitive (or receiver-polyvariant), thus it is more
justified to keep it as an outstanding environment value. Otherwise,
we just need more zig-zags in the code.

> If we generalize to allow variables/parameters other than outer to be non-hot, then the temperature of outer can be recorded together with the temperature of other variables/parameters, rather than on thisV.

This does not sound like a good idea from the engineering point of
view.  The are many equivalent ways to view things abstractly, but
some representations are more natural and simpler than others. To the
extreme, we can also eliminate the concept of `object` in OOPL
semantics with environment and closures, but there is no real
engineering benefits in doing so.

The fact that we associate outers with object in concrete semantics
shows that it's a natural and convenient representation.

On the other hand, a design consideration of ADI is to be as close as
possible to the concrete semantics in the design. Doing things
otherwise will be inconsistent and reaps no engineering benefit.


> I find Warm a bit confusing because it is not used in general for warm values  (all fields assigned with not necessarily hot values) but only for the special case of a ThisRef whose outer is non-hot.

This is because currently we only support warm objects of inner
classes. We can easily support leaking non-hot values to constructors,
thus have `Warm` represent other values as well. `Warm` is just a key
to the abstraction summary of similarly initialized concrete objects.

> After this, one possible further precision improvement could be to extend the Value lattice with something between Warm and Cold, an object with some fields assigned and others not, much like Objekt, but as a Value, and thus with immutable fields.

It's tempting to allow such use cases. However, we need to be cautious
here, because _what the analysis can do does not imply it should
do_. There are two concerns:

- Whether allowing subtle interference of two objects is a good practice
- It has a big impact on worse-case performance (care also needs to be taken for termination)

Coming back to the PR itself, this PR is intended to achieve the following
engineering benefits:

- Enable caching for outers of warm objects and simplify the logic
- Uniformly handle `Warm` and `ThisRef` in semantics
- Make it easier to handle warm objects where the class parameters can be non-hot
- Use opaque types for some abstract domains

I believe the current PR is the simplest from the perspective of
engineering and it paves the way for future extension. It benefits a
lot from our discussions --- see how far we have been from the
type-and-effect system. That said, I'm always open to better
proposals that enjoy the same engineering qualities.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 1]
```diff
@@ -65,43 +65,89 @@ class Semantic {
...
 
   /** Abstract heap stores abstract objects
    *
    *  As in the OOPSLA paper, the abstract heap is monotonistic.
    *
    *  This is only one object we need to care about, hence it's just `Objekt`.
    */
-  type Heap = Objekt
-  def heap(using h: Heap): Heap = h
+  object Heap {
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Before this change, the `Heap` contained only one object, the object at whose constructor we begin the analysis with `init`.

After this change, the `Heap` contains more than one object. Which objects are they? How do we decide which concrete objects are represented in the `Heap` and which concrete objects are represented just by abstract values like `Hot` or `Cold`? What is the intended concretization of `Warm(C, outer)` as an `Addr`, which concrete object(s) does it represent? Does a `Heap` always contain exactly one `ThisRef` or can it contain multiple `ThisRef`s with different values of `klass`? If it can contain multiple `ThisRef`s, what is their concretization?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Thank you @olhotak. These are good theoretical concerns. Indeed, the refactoring blurs the theoretical structure of the analysis.

I claimed that in essence this PR is the same as master, but I was not clear on why they are the same, and how we arrive at the refactoring. Inspired by your questions above, I think we can see the refactoring in steps, which should answer the questions above:

**1. Introduce Summary for `Warm`**

First, to enable caching of outers for warn objects (and simplify the logic in `resolveThis`), we would like to store the immediate outers of all super-classes in a map, let's call it _summary_:

``` Scala
type Summary = Map[Warm, Objekt]
```

Note that, in the above we use `Objekt` that caches the fields in addition to the outers. The rationale is, why not go a step further to cache fields as well? This way, `Value.select` can avoid dealing with the low-level `cache` for expressions. And in the future, if class parameters can be non-hot, they can easily be handled by storing them as fields.

At this point, we still only have one object in the heap, the _summary_ and _heap_ are two different entities. In essence, it's no different from the master. 

**2. Avoid code duplication**

As you insightfully pointed out in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509, we should avoid the code duplication in `Value.select` and `Value.call` for `ThisRef` and `Warm`.

As `Warm` is a key to a map, to unify them, we would make `ThisRef` a key to the same map as well:

``` Scala
trait Key extends Value
case class ThisRef(...) extends Key
case class Warm(...) extends Key

type Summary = Map[Key, Objekt]
```

To make sure that the heap only contains on `ThisRef` as key, we remove `ThisRef` from the summary after checking each class. This refactoring would still be the safe as the master.

But why bother removing the key from the summary, if the key is never reachable? So it's fine to keep multiple `ThisRef` keys in the summary -- conceptually we may think the unreachable key is garbage collected.

However, there is a small semantic different between `ThisRef` and `Warm`: selecting a missing field in `ThisRef` is an error, while the field should be computed on-demand for `Warm`. That's the only specialized logic we need.

This overloading usage of summary is directly inspired by your comment in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509. Overloading of store-like structures happens not only in engineering, but also exists in theoretical work. For example, in [_Abstracting abstract machines_][aam], it uses `Store` to hold both continuations and closures (Section 2.3).

[aam]: https://dl.acm.org/doi/10.1145/1863543.1863553

**3. Renaming**

As the summary contains abstract objects: 

- `ThisRef` refers to objects which may have uninitialized fields
- `Warm` refers to objects whose fields are all initialized but may reach objects with empty fields

Why not rename `key` as `Addr` and `Summary` as `Heap`, which corresponds concrete semantics? Finishing this step leads to the current PR:

``` Scala
trait Addr extends Value
case class ThisRef(...) extends Addr
case class Warm(...) extends Addr

type Heap = Map[Addr, Objekt]
```

This PR also use opaque types for `Heap` instead of type aliases. But such refactoring is cosmetic.

To summarize, the refactoring is inspired by your comments in https://github.com/lampepfl/dotty/pull/12495#discussion_r634745509 and benefited a lot from our previous discussions and the insights from the OOPSLA paper (in OOPSLA we do `this` substitution instead of having cache for `Warm`). The refactoring might have made the theoretical structure less clear, but from the engineering perspective, it seems to be a win.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: I'll try to explain the fundamental problem. `ThisRef` and `Warm` are abstractions that represent some number of concrete objects. If an `Addr` can represent more than one concrete object, then it is unsound to call `updateField` or `updateOuter` on that `Addr`. I understand that `ThisRef(C)` represents a unique concrete object, so that's fine. But we need to be sure that `Warm(C, outer)` also represents a unique concrete object before we can use it as an `Addr`. This is why I am asking *which* concrete object `Warm(C, outer)` represents.

My understanding of `Warm(C, outer)` is that it represents *all* objects of class `C` whose fields have all been assigned and whose `outer` has the given abstract value. If this understanding is correct, then this change introduces unsoundness, because there can be multiple such concrete objects. If this understanding is not correct, then I would like to correct it by identifying the unique concrete object that `Warm(C, outer)` represents.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Would it be correct to understand that `Warm(C, outer)`, when used as an `Addr` for the `Heap`, does not represent *all* warm objects of class `C`, but only *the* unique object on which we invoke `init` to analyze its constructor, in the case when that object is not hot because its outer is not hot?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > My understanding of Warm(C, outer) is that it represents all objects of class C whose fields have all been assigned and whose outer has the given abstract value. If this understanding is correct, then this change introduces unsoundness, because there can be multiple such concrete objects. If this understanding is not correct, then I would like to correct it by identifying the unique concrete object that Warm(C, outer) represents.

To be short, it's not unique and I believe it's sound. I don't think being unique is a necessary requirement in all analyses, and definitely not in our case. One example is type analysis, where we abstract the type of a field by the type of the initializer. Then we reject the program if later a value of a different type is assigned to the field.

The same holds for our analysis, the field abstractions cannot be changed by mutation, and it's uniquely determined by the specific value of `Warm`. The `updateField` and `updateOuter` are in essence cache operations --- the same values can be computed with `eval` or fetched from the low-level expression-based cache. All those values are determined by a specific choice of `Warm`. As all objects that are abstracted by the same `Warm` have the same abstract field values, it is safe to compute, cache, and reuse them. 

From anther perspective, this analysis is equivalent to the OOPSLA type-and-effect system (but with better engineering qualities). Any unsoundness example will in turn be a counter-example to the OOPSLA paper.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > Would it be correct to understand that `Warm(C, outer)`, when used as an `Addr` for the `Heap`, does not represent _all_ warm objects of class `C`, but only _the_ unique object on which we invoke `init` to analyze its constructor, in the case when that object is not hot because its outer is not hot?

`Warm(C, outer)` is an abstraction for all objects of `C` with the same `outer` at a specific time point during execution of the program. For example, given the following program:

```Scala
class C {
    class B() { ... }
    val a = new B
    val b = new B
    // here
}
```

There are at least two objects above abstracted by `Warm(C, ThisRef(C))` after field `b`. The reason why we can abstract with the same warm value is that they are indistinguishable in the abstract domain --- all field values and outers agree.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Based on the discussion above, would it be correct to state the following invariants?

1. The analysis never queries `heap(ThisRef(D))` where `D` is any class other than the class on which `init` was launched. Although the `heap` may contain such `ThisRef`s, they are stale and it would be an error to ask about them.
2. The analysis never queries `heap(ThisRef(C)).outers`. It would be an error to do so.
3. `heap(Warm(C,outer)).field(s) == eval(s.defTree.rhs, Warm(C, outer), C)`
4. `heap(Warm(C,outer)).outers(D) == outerValue(?, Warm(C, outer), D, ?)`

In other words, `heap(ThisRef(C))` maintains the state of the object on which `init` was launched, but `heap(Warm)` is only a cache of `eval` and `outerValue`.

In that case, 3 and 4 are non-trivial results and should have at least some informal proof (not necessarily in the code, of course, but if these are lemmas to be proven externally, they should at least be stated in comments in the code). In particular, are 3 and 4 preserved if `outer` is `Warm` but we then to early promotion of it?

It would also be helpful to list, somewhere, what all the possible values of `outer` in a `Warm` are. Are all `Value`s possible, or only `Warm` and `ThisRef(C)` where `C` is the class on which `init` was launched?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > 1. The analysis never queries heap(ThisRef(D)) where D is any class other than the class on which init was launched. Although the heap may contain such ThisRefs, they are stale and it would be an error to ask about them.

Yes, this is correct.

> 2. The analysis never queries heap(ThisRef(C)).outers. It would be an error to do so.

There is flexibility here: either we handle outers for `ThisRef` uniformly as `Warm`, in spite of knowing that they can only   be `Hot`. Or, in `resolveThis`, we have a special case for `ThisRef`. Both works well. Current PR uses the latter. But for uniformity, it's better to do the former.

> 3. heap(Warm(C,outer)).field(s) == eval(s.defTree.rhs, Warm(C, outer), C)

Yes, that's how we compute the value for fields of warm objects -- the place where we call `updateField`. Thus it holds by construction.

> 4. heap(Warm(C,outer)).outers(D) == outerValue(?, Warm(C, outer), D, ?)

Yes, that's how we compute the outers for warm objects  -- the place where we call `updateOuter`. Thus it holds by construction.

> In other words, heap(ThisRef(C)) maintains the state of the object on which init was launched, but heap(Warm) is only a cache of eval and outerValue.

Yes, that's correct. But `heap(ThisRef(C))` can also be regarded as a cache of the abstract values for the fields, so that later access does not need re-computation.

> In particular, are 3 and 4 preserved if outer is Warm but we then to early promotion of it?

3 and 4 hold by construction, thus they are not affected by early promotion. Early promotion does not change summaries for `Warm`.

> It would also be helpful to list, somewhere, what all the possible values of outer in a Warm are. Are all Values possible, or only Warm and ThisRef(C) where C is the class on which init was launched?

That's a good suggestion to improve the documentation of `Warm`. The implicit invariant here is that the `outer` can only be `RefSet`, `Warm` and `Cold`. It should not be `Hot` nor `Fun`.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: For invariants 3 and 4, the concern is not that the summary (the LHS of the invariant) would change, but that the result of `eval` and `outerValue` (the RHS of the invariant) could change after a change to the fields of the object on which `init` was launched. Specifically, what if the `rhs` of a field of an inner class depends somehow on the outer class, so evaluating the `rhs` changes after the outer class becomes more initialized?

For invariant 3, since `eval` already uses `cache`, why do we need to cache its result again in `heap`?

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: I guess what's confusing me is that the behaviour of `eval` depends on how many fields are in `heap(ThisRef(C)).fields`, but it presumably should not depend on how many fields are in `heap(Warm(C,outer)).fields` or on `heap(???).outers`. Understanding the inputs that `eval` depends on is important to understanding how the analysis works. The confusion is that things that `eval` does and does not depend on are mixed together in `heap`.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]:
``` Scala
class C {
    class Base { def m: C }
    class MIsCold extends Base { def m = C.this }
    class MIsHot extends Base { def m = new C }
    class B(base: Base) { val field = base.m }
    val a = new B(new MIsHot)
    val b = new B(new MIsCold)
    // here
    leak(a.field)
    leak(b.field)
    val c = ???
}
```

At `// here`, what is `heap(Warm(B,ThisRef(C))).fields(field)`? For `a` it should be `Hot` but for `b` it should be `Cold`, shouldn't it?

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Or is it the case that although the type declaration makes `fields` and `outers` mutable for all `Addr` in the heap, we
1. mutate them only during the execution of `init`, and
2. during the execution of `init(thisV = addr)` we mutate only `heap(addr)`, and not the heap of any other addresses.
Then after `init(thisV = addr)` finishes, `heap(addr)` is never mutated anymore. We use it only in a read-only manner to determine what the state was at the end of the constructor of the object represented by `addr`.

Is that right?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > For invariants 3 and 4, the concern is not that the summary (the LHS of the invariant) would change, but that the result of `eval` and `outerValue` (the RHS of the invariant) could change after a change to the fields of the object on which `init` was launched. Specifically, what if the `rhs` of a field of an inner class depends somehow on the outer class, so evaluating the `rhs` changes after the outer class becomes more initialized?

This relies on a fundamental idea of the analysis: the changes in the heap don't have any impact on the result of `eval` --- the result of an expression only depends on `thisV` (and `Env` once we have the extension). That's why we can do cache for expressions without having `heap` as part of the key.

That's also the idea in the type-and-effect system where we can create summaries for fields and methods without caring about how many fields are initialized.

> 
> For invariant 3, since `eval` already uses `cache`, why do we need to cache its result again in `heap`?

The reason is a bit hidden in the long comment, the reason is uniformity for engineering benefits, and support future extension to non-hot class parameters:

"Note that, in the above we use Objekt that caches the fields in addition to the outers. The rationale is, why not go a step further to cache fields as well? This way, Value.select can avoid dealing with the low-level cache for expressions. And in the future, if class parameters can be non-hot, they can easily be handled by storing them as fields."

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > I guess what's confusing me is that the behaviour of eval depends on how many fields are in heap(ThisRef(C)).fields, but it presumably should not depend on how many fields are in heap(Warm(C,outer)).fields or on heap(???).outers. Understanding the inputs that eval depends on is important to understanding how the analysis works. The confusion is that things that eval does and does not depend on are mixed together in heap.

Strictly speaking, the result of `eval` does not depend on the heap for a given valid program, neither `heap(ThisRef(C))` nor `heap(Warm(C,outer))`. Otherwise, the cache has to make the heap part of the key --- which is one of the most important design insights. Similarly, in the type-and-effect system, the summary for an expression does not care about the initialization state of `this`.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > ```scala
> class C {
>     class Base { def m: C }
>     class MIsCold extends Base { def m = C.this }
>     class MIsHot extends Base { def m = new C }
>     class B(base: Base) { val field = base.m }
>     val a = new B(new MIsHot)
>     val b = new B(new MIsCold)
>     // here
>     leak(a.field)
>     leak(b.field)
>     val c = ???
> }
> ```
> 
> At `// here`, what is `heap(Warm(B,ThisRef(C))).fields(field)`? For `a` it should be `Hot` but for `b` it should be `Cold`, shouldn't it?

This depends on what abstract domains we are working on. In the current abstract domain, all class parameters must be hot, thus the abstraction for `field` is always `Hot`.

One we extend the domain to allow non-hot class parameters:

```
case class Warm(klass: ClassSymbol, outer: Value, ctor: Symbol, args: List[Value]) extends Addr
```

Then the abstraction for `field` is dependent on the abstract value of `args`. Here we need to widen `args` the same way as we widen `outer` to finitize the addresses to ensure termination, good performance, and usability.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > Or is it the case that although the type declaration makes `fields` and `outers` mutable for all `Addr` in the heap, we
> 
> 1. mutate them only during the execution of `init`, and

Yes, that's the only place where `updateOuter` and `updateField` is called. And because of the opaque types, they cannot be accidentally mutated.

> 2. during the execution of `init(thisV = addr)` we mutate only `heap(addr)`, and not the heap of any other addresses.
>    Then after `init(thisV = addr)` finishes, `heap(addr)` is never mutated anymore. We use it only in a read-only manner to determine what the state was at the end of the constructor of the object represented by `addr`.
> 

Yes, that's correct. It's in that sense it's equivalent to the type-and-effect system. In particular, in an assignment, we don't update the heap. This makes it different from typical analysis, where we do a union on the heap for assignment.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: > This relies on a fundamental idea of the analysis: the changes in the heap don't have any impact on the result of `eval` --- the result of an expression only depends on `thisV` (and `Env` once we have the extension). That's why we can do cache for expressions without having `heap` as part of the key.

I don't understand this point. `eval` calls `Value.select` on values other than `thisV`, and `select` calls `heap(value).fields`.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: I was not clear, by `thisV` I mean the value for `this`, not `ThisRef`.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:108]: Note that all the values stored in the heap depend on `thisV`, so the heap is just a cache or summary.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 2]
```diff
@@ -65,43 +65,89 @@ class Semantic {
...
-  case class Objekt(klass: ClassSymbol, val fields: mutable.Map[Symbol, Value]) {
-    val promotedValues = mutable.Set.empty[Value]
-  }
+  case class Objekt(klass: ClassSymbol, fields: mutable.Map[Symbol, Value], outers: mutable.Map[ClassSymbol, Value])
 
   /** Abstract heap stores abstract objects
    *
    *  As in the OOPSLA paper, the abstract heap is monotonistic.
    *
    *  This is only one object we need to care about, hence it's just `Objekt`.
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:106]: Remove this line of the comment.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 3]
```diff
@@ -231,62 +294,43 @@ class Semantic {
...
+            if !needResolve then
+              meth
+            else if superType.exists then
+              resolveSuper(addr.klass, superType, meth)
             else
-              resolve(thisRef.klass, meth)
+              resolve(addr.klass, meth)
           if target.isOneOf(Flags.Method | Flags.Lazy) then
             if target.hasSource then
+              val cls = target.owner.enclosingClass.asClass
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:316]: This changed from `target.owner.asClass` to `target.owner.enclosingClass.asClass`. Is that intended?

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 4]
```diff
@@ -307,18 +351,19 @@ class Semantic {
...
+        case addr: Addr =>
           // widen the outer to finitize addresses
-          val outer = if warm.outer.isInstanceOf[Warm] then warm.copy(outer = Cold) else warm
+          val outer = addr match
+            case Warm(_, _: Warm) => Cold
+            case _ => addr
+
           val value = Warm(klass, outer)
+          if !heap.contains(value) then
+            val obj = Objekt(klass, fields = mutable.Map.empty, outers = mutable.Map(klass -> outer))
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:365]: Will this `Objekt` continue to have empty `fields` and `outers` forever since we never call `init` on the new `value`?

**Review by olhotak — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala (hunk 4 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:365]: The call happens at line 367 -- it's a constructor call on the object.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala (hunk 3 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:316]: Yes, it's intentional -- here we reuse the code for calling local methods.

The related code change can be found below for `case id: Ident =>`.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Semantic.scala]: Good catch, now it's removed
